### PR TITLE
Docker Configuration for Containerized Access to Firebase

### PR DESCRIPTION
**Pull request by rlnsy**
## Changes
- Creates a dockerized version of the firebase backend emulator. Running the compose stack provides access to the emulated services on local ports.
- Makes changes to firebase configs, setup script, where needed.

## Requirements
Requires: Docker.

## PR Dependencies
N/A

## Merged Branches and Other Dependencies
Merged branch(es): None
Other dependencies: None

## Subtasks / Progress
N/A

## Preview
N/A

## Testing
Manual

**Instructions**:
For those that wish to test this, here are the steps:
1. Get [Docker](https://docs.docker.com/get-docker/) installed (should have the `docker-compose` command available). Please note that docker can sometimes be resource-intensive, but usage can be controlled via the GUI preferences screen.
2. Get a fresh clone of this branch on the same machine where you installed Docker.
3. In the root directory, run
```
docker-compose up
```
This will take quite a while the first time, as images have to be downloaded and our containers built. There might be some warnings in the logs once the `yarn serve` command is run, but this is expected.
4. If everything worked properly, check that the emulator UI is visible at the given URL. NOTE: there is a bug which prevents the firestore UI from loading. This will be addressed in a later patch.
5. When you are done, exit and run `docker-compose down` to stop all containers.

**What did this do?**
The `Dockerfile` has a sequence of steps that when performed on an instance of a linux container, produce a custom image containing our emulated code whose only dependency is the docker client on your machine. This means that we can build a development environment without any additional software, such as firebase tools, and lets us completely define our stack for testing purposes.

### CODE REVIEW

**Comment by cara-wong** [comment]
@rlnsy Hey Rowan, sorry I've never used Docker before - are you able to add some brief info/steps to test this PR?

**Comment by rlnsy** [comment]
Yeah of course! I should have included that anyway.

**Comment by rlnsy** [comment]
@cara-wong @crugwiro instructions have been added!

**Review by cara-wong — COMMENTED** [review]
Works as per the instructions! Screenshot of the emulator UI below. Is the firestore emulator also supposed to be running in addition to functions?

Screen Shot 2021-03-03 at 10 24 59 PM

**Comment by rlnsy** [comment]
@cara-wong Good point. I wasn't able to get the firestore emulator running either. I should probably try to get it working before this is marked as complete.

**Comment by rlnsy** [comment]
@cara-wong @crugwiro I have enabled the capability for firestore. It seems to be working for the purposes of my automated tests, but there's a bug in the way the UI page loads for firestore (you can see some error messages in the browser console). Since the primary purpose of this PR is to enable containerization for testing, I might hold off on debugging that feature. Thoughts?

**Comment by rlnsy** [comment]
Please also see the updated run command above.

**Comment by rlnsy** [comment]
@cara-wong @crugwiro Looks like I broke the command in one of those ninja commits I made. Will issue a fix later today.

**Comment by cara-wong** [comment]
@rlnsy sorry for the delay on actioning on this PR. Once you've fixed the issue I'll add my review today!

**Comment by rlnsy** [comment]
@crugwiro @cara-wong it is fixed. (reverted command to the original one, as the change wasn't as necessary as I thought)

**Review by cara-wong — COMMENTED** [review]
@rlnsy Tested this again as per the instructions at the top - looks good to me! I think if the firestore emulator is working for automated testing it's ok to wait to debug the errors in the browser console, but maybe we should hold off on merging it until then

**Comment by crugwiro** [comment]
@rlnsy just got this working locally, Finally ! I think as Cara mentioned, we should hold off on merging it for now.

**Comment by rlnsy** [comment]
@crugwiro awesome thanks. I'll take a look at fixing the ui errors soon.

**Review by crugwiro — APPROVED** [review]
Tested as per instructions, looks good.

**Comment by rlnsy** [comment]
@wynnset this is ready to merge once the lock update is merged.

File: Dockerfile
Comment (wynnset) [inline Dockerfile:1]: Can we move this under bipolar bridges instead?

**Review by wynnset — COMMENTED** [review]
Just one minor comment but everything else looks good to get merged.

File: Dockerfile
Comment (rlnsy) [reply Dockerfile]: Done.

**Comment by rlnsy** [comment]
Image reference updated. Ready to merge.

**Review by wynnset — APPROVED** [review]

**Pull request merged**
